### ISSUE TITLE
Allow IsByRefLike and IsReadOnly anonymous records

### ISSUE THREAD

**Issue by cartermp** [fsharp/fslang-suggestions#712]
I propose we allow [ByRefLike struct](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/structures#byreflike-structs ) anonymous records.

You can create struct anonymous records as such today:

```fsharp
let f (r: struct {| IntVal: int |}) = r.IntVal
f {| IntVal=12 |}
// Or if you don't like structness inference: f struct {| IntVal=12 |}
```

However, there is no ability to make them ByRefLike, which means you cannot have them contain other ByRefLike structs such as [`Span<'T>`](https://docs.microsoft.com/en-us/dotnet/api/system.span-1?view=netcore-2.2).

A possible declaration syntax might look like this:

```fsharp
// Using an attribute
let f (r: [<IsByRefLike>] struct {| S: Span<int> |}) = S.Length

// Using a keyword
let a (r: ref struct {| S: Span<int> |}) = S.Length
let b (r: struct ref {| S: Span<int> |}) = S.Length
let c (r: byref struct {| S: Span<int> |}) = S.Length
let d (r: struct byref {| S: Span<int> |}) = S.Length
let e (r: byreflike {| S: Span<int> |}) = S.Length // 'byrefLike' implies struct
```

It would be sort of nonsense to have you require an attribute to instantiate one, so a keyword would likely be needed:

```fsharp
a ref struct {| S = Span<int>.Empty |}
b struct ref {| S = Span<int>.Empty |}
c byref struct {| S = Span<int>.Empty |}
d struct byref {| S = Span<int>.Empty |}
e byreflike {| S = Span<int>.Empty |} // 'byrefLike' implies struct
```

Or perhaps structness inference would take care of the instantiation site:

```fsharp
// Inferred to by a ByRefLike struct anonymous record
a {| S = Span<int>.Empty |}
```

Similarly, there is no way to define a `IsReadOnly` struct anonymous record. Possible declaration syntax:

```fsharp
// Using an attribute
let f (r: [<IsReadOnly>] struct {| S: Span<int> |}) = S.Length

// Using a keyword
let a (r: readonly struct {| S: Span<int> |}) = S.Length
let b (r: struct readonly {| S: Span<int> |}) = S.Length
let c (r: readonly {| S: Span<int> |}) = S.Length // 'readonly' implies struct
```

And an instantiation syntax:

```fsharp
a readonly struct {| S = Span<int>.Empty |}
b struct readonly {| S = Span<int>.Empty |}
c readonly {| S = Span<int>.Empty |} // 'byrefLike' implies struct
```

## Pros and Cons 

Advantages:

* Parity with records
* Allows you to use these more ephemeral constructs in the perf-sensitive environments ByRefLike structs are intended for

Disadvantages:

* I can't think of an explicit instantiation syntax that doesn't suck
* This can make an input parameter heavy syntax-wise
* Allowing an attribute now requires both an attribute _and_ a keyword to define something
* Disallowing an attribute would feel odd considering it's how you define something to by ByRefLike today in F#
* Allowing a keyword instead of an attribute is inconsistent with how you define ByRefLike structs today in F#

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): **M**

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by dsyme** [comment]
I don't think this looks essential, given the overall design goals for anon records
